### PR TITLE
Improved code formatter

### PR DESCRIPTION
**Pull request by sebmck**
This PR rewrites the Rome JS code formatter. Rather than constantly pushing onto a single buffer as we generate the AST, we know built up "tokens" which signify different chunks of the output. These tokens can be rich, indicating if they contain a break point, and information to print things like lists. This allows us to do much better line wrapping.

Some of the current formatting isn't ideal, but with this architecture we can improve it in subsequent PRs. This branch has gotten very unwieldly and it's hard to follow exactly what's happening so I'd like to get this merged ASAP.

## Example

This is an example of the generator for `ConditionalExpression`:

```
return [
  ...generator.print(node.test, node),
  space,
  group([
    [operator('?'), space, ...generator.print(node.consequent, node)],
    [operator(':'), space, ...generator.print(node.alternate, node)],
  ], {
    priority: true,
    broken: {
      before: [newline],
      indentNewline: false,
      separator: [newline],
    },
    unbroken: {
      separator: [space],
    },
  }),
];
```

There's a few things going on here. First we're generating the test node. This returns an array of tokens. We then spread it in, and insert a space token. Then, we declare a group. A group is a boundary where lines can be broken. It takes an array of tokens and loops over them deciding what tokens to output before, after and between them depending on the group options.

In this instance, we're declaring that to break the group, insert a newline between all items. When the group is unbroken we should just use a space.

By default when all groups are broken they're indented. You can customize this with the `broken.indent` property. Accompanying the indentation is also a newline before and after. In this instance we don't want the newline after so we disable the indentation newlines with `broken.indentNewline: false`.

We set `priority: true` on this group to tell nested groups, let us break first before you do. So if you have something like an object in the `alternate` clause, we wont break it until we've first broken the conditional up.

### LinkedGroup

There are times though when you want to break a group when you're a sibling. For example:

```
const makePerson = (name: string): Person => {};
```

In this instance, the binding parameters are a group. But what if the return type is too long and needs to be broken? The nicest way would be to first break the params, but we are no longer in that group. You can use `linkedGroup` for this. It takes an array of tokens that could contain a group. It then assigns a boundary that takes priority over all within it. When the linked group is broken, it breaks the first unbroken group it finds. If then that doesn't satisfy line wrapping then it will move onto the next child group etc.

## Pending work

Source maps aren't currently supported in this PR. Once I add support for them and clean up `Generator.ts` then we can merge.

## Improvements/quirks

**Array allocation.** We do a lot of array allocation. I'm not sure if there's a way to avoid a lot of these. @ooflorent also mentioned that creating a new array and spreading may be worse than just `push`ing onto it. I don't want to prematurely optimize, and I don't think we should spend a ton of time benchmarking this until we need to, but we should just go with whatever option is the most ergnomic, which is probably using `push` and `concat`.

**Throwing errors.** Right now we are throwing errors to signal to callers that a line has been exceeded. The group printer will then catch it and reprint itself as broken. I believe this is actually "algebraic effects" (lol). Throwing errors used to be expensive in JS, maybe it still is, but I'm not sure it would be any faster to implement some other magic to handle it.

**Some formatting is subpar.** It's better than the old code generator though. Now that we have the abstractions in place we'll be able to make more incremental improvements.

### CODE REVIEW

**Comment by sebmck** [comment]
This PR also contains reformatting all the files. Pay attention to only the first commit if you're interested in the functional changes.

**Comment by sebmck** [comment]
I also deliberately did not look into how prettier works, although I vaguely know it follows a similar model. I like going into problems without any reference as we may have different constraints than others or discover more efficient solutions.

**Comment by ooflorent** [comment]
For documentation, here are a few weird formatting:

```js
// current
let contextEndIndex = coerce0(Math.min(
  allLines.length - 1,
    get0(endLineIndex) +
    CODE_FRAME_CONTEXT_LINES,
));

// expected
let contextEndIndex = coerce0(Math.min(
  allLines.length - 1,
  get0(endLineIndex) + CODE_FRAME_CONTEXT_LINES,
));
```

```js
// current
const noMarkerLine = get0(markerOffset) === 0 && pointerLength === 1 &&
    markerMessage ===
    '';

// expected (test should not break)
const noMarkerLine = get0(markerOffset) === 0 && pointerLength === 1 &&
  markerMessage === '';
```

```js
// current
export function parseCLIFlags<
  T
>(opts: ParserOptions<T>): ParserInterface<T> {

// expected (there no reason to break this line)
export function parseCLIFlags<T>(opts: ParserOptions<T>): ParserInterface<T> {

// expected (break params before types)
export function parseCLIFlags<T>(
  opts: ParserOptions<T>
): ParserInterface<T> {
```

```js
// current
throw new Error(
    "Trying to setStreamColumns on a stream that isn't attached to this Reporter",
  );

// expected (weird indentation)
throw new Error(
  "Trying to setStreamColumns on a stream that isn't attached to this Reporter",
);
```

```js
// current
unexpected(
  {
    message: `The character <emphasis>${escapeMarkup(char)}</emphasis> isn't allowed`,
    start: add(offset, i),
  },
);

// expected (calls with single object should not be indented)
unexpected({
  message: `The character <emphasis>${escapeMarkup(char)}</emphasis> isn't allowed`,
  start: add(offset, i),
});
```

```js
// current
if (typeof value === 'function' || typeof value === 'undefined' ||
      typeof value ===
      'symbol') {

// expected (test should be break)
if (typeof value === 'function' || typeof value === 'undefined' ||
    typeof value === 'symbol') {

// expected (improve readability)
if (typeof value === 'function' ||
    typeof value === 'undefined' ||
    typeof value === 'symbol') {
```

**Comment by ooflorent** [comment]
This change makes a huge difference:

```diff
--- a/packages/@romejs/js-formatter/builders/expressions/AssignmentExpression.ts
+++ b/packages/@romejs/js-formatter/builders/expressions/AssignmentExpression.ts
@@ -49,7 +49,7 @@ export default function AssignmentExpression(

   const right = builder.tokenize(node.right, node);

-  tokens.push(breakGroup([[...left, space, sep], right]));
+  tokens.push(breakGroup([[...left, space, sep], right], true));

   if (needsExtraParens) {
     tokens.push(operator(')'));
```

It fixes most of the awkward breaks. However, the indentation is not properly handled. It think it's worth pursuing the refactor including this change.

The motivation behind this change is to fix this case:

```js
if (x === y && y ===
    z) {
```

After the change, the above snippet would be:

```js
if (x === y &&
    y === z) {
```

**Comment by ooflorent** [comment]
Another one that plays well:

```diff
--- a/packages/@romejs/js-formatter/Builder.ts
+++ b/packages/@romejs/js-formatter/Builder.ts
@@ -208,11 +208,13 @@ export default class Builder {
       forceBroken?: boolean;
       indentNewline?: boolean;
       indent?: boolean;
+      priority?: boolean;
     } = {},
   ): GroupToken {
     return this.tokenizeJoin(items, parent, {
       breakOnNewline: opts.breakOnNewline,
       newline: true,
+      priority: opts.priority,
       broken: {
         force: opts.forceBroken,
         indent: opts.indent,
--- a/packages/@romejs/js-formatter/builders/expressions/CallExpression.ts
+++ b/packages/@romejs/js-formatter/builders/expressions/CallExpression.ts
@@ -29,6 +29,7 @@ export default function CallExpression(builder: Builder, node: AnyNode): Tokens
     operator('('),
     builder.tokenizeCommaList(node.arguments, node, {
       trailing: true,
+      priority: true,
     }),
     operator(')'),
   ];
```

Combined with the previous suggested change, the readability greatly improves:

```diff
@@ -157,9 +156,8 @@ export default function buildMessageCodeFrame(
   // If what the marker is highlighting equals the marker message then it's redundant so don't show the message
   if (markerMessage !== '') {
     const text = sourceText.slice(get0(start.index), get0(end.index));
-    if (cleanEquivalentString(text) === cleanEquivalentString(stripMarkupTags(
-        markerMessage,
-      ))) {
+    if (cleanEquivalentString(text) ===
+        cleanEquivalentString(stripMarkupTags(markerMessage))) {
       markerMessage = '';
     }
   }
```

```diff
@@ -30,9 +30,12 @@ export default function highlightCode(opts: AnsiHighlightOptions): string {
   if (opts.language === 'js') {
     // js-parser does not accept an "unknown" sourceType
-    return ansiHighlightJS(opts.input, opts.sourceType === undefined ||
-        opts.sourceType ===
-        'unknown' ? 'script' : opts.sourceType);
+    return ansiHighlightJS(
+      opts.input,
+      opts.sourceType === undefined || opts.sourceType === 'unknown'
+        ? 'script'
+        : opts.sourceType,
+    );
   }
```

```diff
@@ -332,10 +338,11 @@ export default async function cli() {
   if (res.type === 'SUCCESS') {
     // Write markers if we were collecting them
     if (requestFlags.collectMarkers) {
-      const markersPath = clientFlags.cwd.resolve(cliFlags.markersPath ===
-        undefined
+      const markersPath = clientFlags.cwd.resolve(
+        cliFlags.markersPath === undefined
           ? `Markers-${getFilenameTimestamp()}.json`
-        : cliFlags.markersPath);
+          : cliFlags.markersPath,
+      );
```

**Comment by sebmck** [comment]
@ooflorent Sweet! Thanks for the suggestions. I applied some of them locally and noticed an improvement in some areas but I think it'd require a bit more tweaking to make it perfect. This PR passes CI so I'll merge it and so feel free to send PRs with any changes since we'll have a better idea seeing the diffs for the whole repo on individual changes.
